validate.go

```diff
@@ -9,87 +9,138 @@ import (
 	"sync"
 	"time"
 
+	"github.com/fatih/color"
+
 	"github.com/aelsabbahy/goss/outputs"
 	"github.com/aelsabbahy/goss/resource"
 	"github.com/aelsabbahy/goss/system"
 	"github.com/aelsabbahy/goss/util"
-	"github.com/fatih/color"
-	"github.com/urfave/cli"
 )
 
-func getGossConfig(c *cli.Context) GossConfig {
+// RuntimeConfig is configuration for various aspects of the goss system and
+// is modeled on the cli context used in the cli tooling.  Mainly the translation
+// into this format is there to support making the system usable as a package
+type RuntimeConfig struct {
+	FormatOptions     []string
+	Vars              string
+	VarsInline        string
+	Spec              string
+	Sleep             time.Duration
+	RetryTimeout      time.Duration
+	Cache             time.Duration
+	Timeout           time.Duration
+	MaxConcurrent     int
+	NoColor           *bool
+	OutputFormat      string
+	PackageManager    string
+	Endpoint          string
+	ListenAddress     string
+	ExcludeAttributes []string
+	Insecure          bool
+	NoFollowRedirects bool
+	Server            string
+	Username          string
+	Password          string
+	Debug             bool
+}
+
+func getGossConfig(vars string, varsInline string, specFile string) (cfg *GossConfig, err error) {
 	// handle stdin
 	var fh *os.File
 	var path, source string
 	var gossConfig GossConfig
-	varsFile := c.GlobalString("vars")
-	varsInline := c.GlobalString("vars-inline")
-	currentTemplateFilter = NewTemplateFilter(varsFile, varsInline)
-	specFile := c.GlobalString("gossfile")
+
+	currentTemplateFilter, err = NewTemplateFilter(vars, varsInline)
+	if err != nil {
+		return nil, err
+	}
+
 	if specFile == "-" {
 		source = "STDIN"
 		fh = os.Stdin
 		data, err := ioutil.ReadAll(fh)
 		if err != nil {
-			fmt.Printf("Error: %v\n", err)
-			os.Exit(1)
+			return nil, err
+		}
+		outStoreFormat, err = getStoreFormatFromData(data)
+		if err != nil {
+			return nil, err
+		}
+
+		gossConfig, err = ReadJSONData(data, true)
+		if err != nil {
+			return nil, err
 		}
-		outStoreFormat = getStoreFormatFromData(data)
-		gossConfig = ReadJSONData(data, true)
 	} else {
 		source = specFile
 		path = filepath.Dir(specFile)
-		outStoreFormat = getStoreFormatFromFileName(specFile)
-		gossConfig = ReadJSON(specFile)
+		outStoreFormat, err = getStoreFormatFromFileName(specFile)
+		if err != nil {
+			return nil, err
+		}
+
+		gossConfig, err = ReadJSON(specFile)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	gossConfig = mergeJSONData(gossConfig, 0, path)
+	gossConfig, err = mergeJSONData(gossConfig, 0, path)
+	if err != nil {
+		return nil, err
+	}
 
 	if len(gossConfig.Resources()) == 0 {
-		fmt.Printf("Error: found 0 tests, source: %v\n", source)
-		os.Exit(1)
+		return nil, fmt.Errorf("found 0 tests, source: %v", source)
 	}
-	return gossConfig
+
+	return &gossConfig, nil
```

Comment:
curious, why change this to a pointer?

Reply:
It’s easier to return nil on error. Also NewGossConfig does the same so I made this do that too. Can revert if you like.

In general I think returning nil, err vs empty config that will anyway just be discarded is more idiomatic - old code just exit’d so seemed like a sane change.

Reply:
Makes sense